Readme.md
# EDK II Expression Parser

This library provides a simple parser that can be used to evaluate the value of an expression which aligns [EDK II Meta-Data Expression Syntax Specification] revision 1.3.

[EDK II Meta-Data Expression Syntax Specification]: https://tianocore-docs.github.io/edk2-MetaDataExpressionSyntaxSpecification/release-1.30/


## Installation

This library requires Python 3.10 or later, and [Pygments].

[Pygments]: https://pygments.org/

This library is intended to be used as part of another project, and I will not make it available on PyPI.

You can use the `pip` to install it from source:

```bash
pip install git+https://github.com/tzing/edk2-expression.git
```

or simply copy the source code into your project.


## Usage

First, parse the expression text into abstract syntax tree (AST):

```python
>>> import edk2_expression
>>> expr = edk2_expression.parse("$(FOO) +2")
```

Then, evaluate the expression with a dictionary of variables:

```python
>>> print(expr.evaluate({"FOO": 1}))
3
```

For nested expression, one can set `nest` to `evaluate` to evaluate it:

```python
>>> expr1 = edk2_expression.parse("True ? $(BAR) + 3 : 0")
>>> print(expr1.evaluate({"FOO": 1, "BAR": expr}, nest="evaluate"))
6
```

See [example/](./example/) directory for more examples.


## Supported Syntax

### Data Fields

| EDK II Data Field Name | Supported      | Evaluate                     |
| ---------------------- | -------------- | ---------------------------- |
| `<Array>`              | No[^array]     | No                           |
| `<CName>`              | Yes            | No                           |
| `<Function>`           | No[^func]      | No                           |
| `<GuidValue>`          | Partial[^guid] | Yes; Returns `uuid.UUID`     |
| `<MacroVal>`           | Yes            | Yes; Returns the macro value |
| `<Number>`             | Yes            | Yes; Returns the number      |
| `<PcdName>`            | Yes            | No                           |
| `<StringLiteral>`      | Yes            | Yes; Returns the string      |
| `<TrueFalse>`          | Yes            | Yes; Returns `True`/`False`  |

[^array]: This library currently reads the array as a string, and does not support any operations or evaluation.

[^func]: Per addressed by EDK II:

    > Functions should only be used if all tools that process the entry in the meta-data file comprehend the function syntax.

    Since this library is designed for general purpose, it does not parse functions.

[^guid]: GUID in `<CformatGuid>` format is currently not parsed.


### Operators

| EDK II Expression Name    | Operator                                     | Supported | Precedence[^prec] |
| ------------------------- | -------------------------------------------- | --------- | ----------------: |
| `<UnaryExpression>`       | `!`, `NOT`, `not`, `~`                       | Yes       |                 2 |
| `<MultiplicativeExpress>` | `*`, `/`, `%`                                | Yes       |                 3 |
| `<AdditiveExpress>`       | `+`, `-`                                     | Yes       |                 4 |
| `<ShiftExpression>`       | `<<`, `>>`                                   | Yes       |                 5 |
| `<RelationalExpress>`     | `<`, `LT`, `<=`, `LE`, `>`, `GT`, `>=`, `GE` | Yes       |                 6 |
| `<EqualityExpression>`    | `==`, `EQ`, `!=`, `NE`                       | Yes       |                 7 |
| `<BitwiseAndExpression>`  | `&`                                          | Yes       |                 8 |
| `<BitwiseXorExpress>`     | `^`                                          | Yes       |                 9 |
| `<BitwiseOrExpress>`      | `\|`                                         | Yes       |                10 |
| `<LogicalAndExpress>`     | `&&`, `AND`, `and`                           | Yes       |                11 |
| `<LogicalXorExpress>`     | `XOR`, `xor`                                 | Yes       |                12 |
| `<LogicalOrExpress>`      | `\|\|`, `OR`, `or`                           | Yes       |                13 |
| `<CondExpress>`           | `?:` (ternary conditional)                   | Yes       |                15 |

[^prec]: The lower the number, the higher the precedence.


## Extras

### Pygments lexer

This parser utilizes [Pygments] for expression text tokenization and, as a result, it comes packaged with a lexer.

If you are using Pygments, you can use the lexer directly:

```python
import edk2_expression.lex

lexer = edk2_expression.lex.Edk2ExpressionLexer()
```

### `DEFINED` function

A function that returns `True` if the macro is defined, otherwise `False`.

The syntax is defined as follows in EBNF:

```ebnf
<MacroDefined> ::= "DEFINED(" <CName> ")"
<CName>        ::= (a-zA-Z_) [(a-zA-Z0-9_)]*
```

This function is not defined in EDK II specification therefore it is disabled by default.
To enable it, uncomment the lines for `<MacroDefined>` in [lexer](./edk2_expression/lex.py):

<details>
<summary>patch</summary>

```diff
diff --git a/edk2_expression/lex.py b/edk2_expression/lex.py
index ac66120..e87340b 100644
--- a/edk2_expression/lex.py
+++ b/edk2_expression/lex.py
@@ -64,10 +64,10 @@ class Edk2ExpressionLexer(RegexLexer):
                 r"(\$\()([A-Z][A-Z0-9_]*)(\))",
                 bygroups(Keyword.Declaration, Name.Variable, Keyword.Declaration),
             ),
-            # (  # <MacroDefined>
-            #     rf"\b(DEFINED)(\()({CName})(\))",
-            #     bygroups(Keyword.Type, Punctuation, Name.Variable, Punctuation),
-            # ),
+            (  # <MacroDefined>
+                rf"\b(DEFINED)(\()({CName})(\))",
+                bygroups(Keyword.Type, Punctuation, Name.Variable, Punctuation),
+            ),
             (rf"\b{CName}\b", Name.Variable),  # <CName>
         ],
         "array": [
```

</details>

## Changelog

See [Changelog.md](./Changelog.md).

## License

This project is licensed under the terms of the MIT license.


Changelog.md
# Changelog

## 0.2.1 (2023-11-24)

* Fix: Prevent output incomplete expression like `1 : 2`
* Doc: Clarify the operator precedence

## 0.2.0 (2023-11-14)

* Enhance: `NestMethod` now supports case-insensitive matching
* Enhance: Early termination is now supported for Logical AND (`&&`) and Logical OR (`||`) during value evaluation

## 0.1.1 (2023-11-03)

* Enhance: Do not raise error on nested constant
* Fix: Adjust the syntax for MacroDefined to match the EBNF in the documentation.

## 0.1.0 (2023-11-01)

First release


tests/test_lex.py
import unittest

from pygments.token import Token

from edk2_expression.lex import Edk2ExpressionLexer


class TestEdk2ExpressionLexer(unittest.TestCase):
    def setUp(self) -> None:
        self.lexer = Edk2ExpressionLexer()

    def lex(self, text: str) -> list[tuple[Token, str]]:
        tokens = []
        for _, token, text in self.lexer.get_tokens_unprocessed(text):
            tokens.append((token, text))
        return tokens

    def test_StringLiteral(self):
        self.assertEqual(
            self.lex("L'SampleString'"),
            [
                (Token.Punctuation, "L'"),
                (Token.String, "SampleString"),
                (Token.Punctuation, "'"),
            ],
        )
        self.assertEqual(
            self.lex('"AnotherString"'),
            [
                (Token.Punctuation, '"'),
                (Token.String, "AnotherString"),
                (Token.Punctuation, '"'),
            ],
        )

    def test_TrueFalse(self):
        self.assertEqual(
            self.lex("TRUE"),
            [
                (Token.Keyword.Constant, "TRUE"),
            ],
        )
        self.assertEqual(
            self.lex("False"),
            [
                (Token.Keyword.Constant, "False"),
            ],
        )

    def test_HexNumber(self):
        self.assertEqual(
            self.lex("0XFF"),
            [
                (Token.Number.Hex, "0XFF"),
            ],
        )

    def test_Integer(self):
        self.assertEqual(
            self.lex("12345"),
            [
                (Token.Number.Integer, "12345"),
            ],
        )
        self.assertEqual(
            self.lex("6"),
            [
                (Token.Number.Integer, "6"),
            ],
        )

    def test_PcdName(self):
        self.assertEqual(
            self.lex("gUefiCpuPkgTokenSpaceGuid.PcdCpuLocalApicBaseAddress"),
            [
                (
                    Token.Name.Entity,
                    "gUefiCpuPkgTokenSpaceGuid.PcdCpuLocalApicBaseAddress",
                ),
            ],
        )

    def test_GuidValue(self):
        self.assertEqual(
            self.lex("123e4567-e89b-12d3-a456-426655440000"),
            [
                (Token.Name.Entity, "123e4567-e89b-12d3-a456-426655440000"),
            ],
        )
        self.assertEqual(
            self.lex(
                "{0x123e4567, 0xe89b, 0x12d3, {0xa4, 0x56, 0x42, 0x66, 0x55, 0x44, 0x00, 0x00}}"
            ),
            [
                (Token.Punctuation, "{"),
                (Token.Literal.Number.Hex, "0x123e4567"),
                (Token.Punctuation, ","),
                (Token.Text.Whitespace, " "),
                (Token.Literal.Number.Hex, "0xe89b"),
                (Token.Punctuation, ","),
                (Token.Text.Whitespace, " "),
                (Token.Literal.Number.Hex, "0x12d3"),
                (Token.Punctuation, ","),
                (Token.Text.Whitespace, " "),
                (Token.Punctuation, "{"),
                (Token.Literal.Number.Hex, "0xa4"),
                (Token.Punctuation, ","),
                (Token.Text.Whitespace, " "),
                (Token.Literal.Number.Hex, "0x56"),
                (Token.Punctuation, ","),
                (Token.Text.Whitespace, " "),
                (Token.Literal.Number.Hex, "0x42"),
                (Token.Punctuation, ","),
                (Token.Text.Whitespace, " "),
                (Token.Literal.Number.Hex, "0x66"),
                (Token.Punctuation, ","),
                (Token.Text.Whitespace, " "),
                (Token.Literal.Number.Hex, "0x55"),
                (Token.Punctuation, ","),
                (Token.Text.Whitespace, " "),
                (Token.Literal.Number.Hex, "0x44"),
                (Token.Punctuation, ","),
                (Token.Text.Whitespace, " "),
                (Token.Literal.Number.Hex, "0x00"),
                (Token.Punctuation, ","),
                (Token.Text.Whitespace, " "),
                (Token.Literal.Number.Hex, "0x00"),
                (Token.Punctuation, "}"),
                (Token.Punctuation, "}"),
            ],
        )

    def test_MACROVAL(self):
        self.assertEqual(
            self.lex("$(TEST)"),
            [
                (Token.Keyword.Declaration, "$("),
                (Token.Name.Variable, "TEST"),
                (Token.Keyword.Declaration, ")"),
            ],
        )

    def test_CName(self):
        self.assertEqual(
            self.lex("testName"),
            [
                (Token.Name.Variable, "testName"),
            ],
        )
        self.assertEqual(
            self.lex("notKeyword"),
            [
                (Token.Name.Variable, "notKeyword"),
            ],
        )

    def test_UintMac(self):
        self.assertEqual(
            self.lex("{UINT8(8)}"),
            [
                (Token.Punctuation, "{"),
                (Token.Keyword.Type, "UINT8"),
                (Token.Punctuation, "("),
                (Token.Number.Integer, "8"),
                (Token.Punctuation, ")"),
                (Token.Punctuation, "}"),
            ],
        )

    def test_Label(self):
        self.assertEqual(
            self.lex("{LABEL(testLabel)}"),
            [
                (Token.Punctuation, "{"),
                (Token.Keyword.Type, "LABEL"),
                (Token.Punctuation, "("),
                (Token.Name.Variable, "testLabel"),
                (Token.Punctuation, ")"),
                (Token.Punctuation, "}"),
            ],
        )

    def test_Offset(self):
        self.assertEqual(
            self.lex("{OFFSET_OF(test)}"),
            [
                (Token.Punctuation, "{"),
                (Token.Keyword.Type, "OFFSET_OF"),
                (Token.Punctuation, "("),
                (Token.Name.Variable, "test"),
                (Token.Punctuation, ")"),
                (Token.Punctuation, "}"),
            ],
        )

    def test_DevicePath(self):
        self.assertEqual(
            self.lex('{DEVICE_PATH(".")}'),
            [
                (Token.Punctuation, "{"),
                (Token.Keyword.Type, "DEVICE_PATH"),
                (Token.Punctuation, '("'),
                (Token.String, "."),
                (Token.Punctuation, '")'),
                (Token.Punctuation, "}"),
            ],
        )

    def test_GuidStr(self):
        self.assertEqual(
            self.lex('{GUID("123e4567-e89b-12d3-a456-426655440000")}'),
            [
                (Token.Punctuation, "{"),
                (Token.Keyword.Type, "GUID"),
                (Token.Punctuation, '("'),
                (Token.Name.Entity, "123e4567-e89b-12d3-a456-426655440000"),
                (Token.Punctuation, '")'),
                (Token.Punctuation, "}"),
            ],
        )
        self.assertEqual(
            self.lex("{GUID(test)}"),
            [
                (Token.Punctuation, "{"),
                (Token.Keyword.Type, "GUID"),
                (Token.Punctuation, "("),
                (Token.Name.Variable, "test"),
                (Token.Punctuation, ")"),
                (Token.Punctuation, "}"),
            ],
        )

        ll = self.lex(
            "{GUID({0x123e4567, 0xe89b, 0x12d3, {0xa4, 0x56, 0x42, 0x66, 0x55, 0x44, 0x00, 0x00}})}"
        )
        self.assertEqual(
            ll[:5],
            [
                (Token.Punctuation, "{"),
                (Token.Keyword.Type, "GUID"),
                (Token.Punctuation, "("),
                (Token.Punctuation, "{"),
                (Token.Literal.Number.Hex, "0x123e4567"),
            ],
        )
        self.assertEqual(
            ll[-5:],
            [
                (Token.Literal.Number.Hex, "0x00"),
                (Token.Punctuation, "}"),
                (Token.Punctuation, "}"),
                (Token.Punctuation, ")"),
                (Token.Punctuation, "}"),
            ],
        )

    def test_UnaryOp(self):
        self.assertEqual(
            self.lex("NOT gExample.Test"),
            [
                (Token.Operator, "NOT"),
                (Token.Text.Whitespace, " "),
                (Token.Name.Entity, "gExample.Test"),
            ],
        )
        self.assertEqual(
            self.lex("-16"),
            [
                (Token.Operator, "-"),
                (Token.Number.Integer, "16"),
            ],
        )

    def test_BinaryOp(self):
        self.assertEqual(
            self.lex("1 +2"),
            [
                (Token.Number.Integer, "1"),
                (Token.Text.Whitespace, " "),
                (Token.Operator, "+"),
                (Token.Number.Integer, "2"),
            ],
        )
        self.assertEqual(
            self.lex("True OR FALSE"),
            [
                (Token.Keyword.Constant, "True"),
                (Token.Text.Whitespace, " "),
                (Token.Operator, "OR"),
                (Token.Text.Whitespace, " "),
                (Token.Keyword.Constant, "FALSE"),
            ],
        )

    def test_TernaryOp(self):
        self.assertEqual(
            self.lex("foo ? bar : baz"),
            [
                (Token.Name.Variable, "foo"),
                (Token.Text.Whitespace, " "),
                (Token.Operator, "?"),
                (Token.Text.Whitespace, " "),
                (Token.Name.Variable, "bar"),
                (Token.Text.Whitespace, " "),
                (Token.Operator, ":"),
                (Token.Text.Whitespace, " "),
                (Token.Name.Variable, "baz"),
            ],
        )

    def test_nested(self):
        self.assertEqual(
            self.lex("(!FALSE)"),
            [
                (Token.Punctuation, "("),
                (Token.Operator, "!"),
                (Token.Keyword.Constant, "FALSE"),
                (Token.Punctuation, ")"),
            ],
        )

        self.assertEqual(
            self.lex("$(FOO) > (5-3)"),
            [
                (Token.Keyword.Declaration, "$("),
                (Token.Name.Variable, "FOO"),
                (Token.Keyword.Declaration, ")"),
                (Token.Text.Whitespace, " "),
                (Token.Operator, ">"),
                (Token.Text.Whitespace, " "),
                (Token.Punctuation, "("),
                (Token.Number.Integer, "5"),
                (Token.Operator, "-"),
                (Token.Number.Integer, "3"),
                (Token.Punctuation, ")"),
            ],
        )

    def test_comment(self):
        self.assertEqual(
            self.lex("1 # comment"),
            [
                (Token.Number.Integer, "1"),
                (Token.Text.Whitespace, " "),
                (Token.Comment.Single, "# comment"),
            ],
        )


tests/test_init.py
from unittest import TestCase

import edk2_expression
from edk2_expression.ast import Expression


class Test(TestCase):
    def test_parse(self):
        expr = edk2_expression.parse("$(FOO) > 5")
        self.assertIsInstance(expr, Expression)
        self.assertTrue(expr.evaluate({"FOO": 6}))
        self.assertFalse(expr.evaluate({"FOO": 4}))


tests/ast/test_ast.py
from unittest import TestCase

from pygments.token import Token

from edk2_expression.ast import parse
from edk2_expression.error import ParseError
from edk2_expression.lex import Edk2ExpressionLexer


class TestParseExpression(TestCase):
    def setUp(self) -> None:
        self.lexer = Edk2ExpressionLexer()

    def lex(self, text: str) -> list[tuple[Token, str]]:
        return list(self.lexer.get_tokens_unprocessed(text))

    def test_simple(self):
        tree = parse(self.lex("$(FOO)"))
        self.assertEqual(str(tree), "$(FOO)")
        self.assertEqual(tree.evaluate({"FOO": 16}), 16)

    def test_unary(self):
        tree = parse(self.lex("not TRUE"))
        self.assertEqual(str(tree), "!True")
        self.assertEqual(tree.evaluate({}), False)

    def test_binary(self):
        tree = parse(self.lex("2 <3"))
        self.assertEqual(str(tree), "2 < 3")
        self.assertEqual(tree.evaluate({}), True)

    def test_ternary(self):
        tree = parse(self.lex("true ? 5:7"))
        self.assertEqual(str(tree), "True ? 5 : 7")
        self.assertEqual(tree.evaluate({}), 5)

    def test_nested(self):
        tree = parse(self.lex("$(FOO) == 5 ? ($(BAR) +3) :$(BAZ)*3"))
        self.assertEqual(str(tree), "$(FOO) == 5 ? ($(BAR) + 3) : ($(BAZ) * 3)")

        self.assertEqual(tree.evaluate({"FOO": 5, "BAR": 3}), 6)
        self.assertEqual(tree.evaluate({"FOO": 3, "BAZ": 7}), 21)

    def test_unbalance_bracket(self):
        with self.assertRaises(ParseError) as cm:
            parse(self.lex("3 + 5)"))
        self.assertEqual(
            str(cm.exception), "Unbalanced brackets in expression '3 + 5)'"
        )

    def test_missing_operand(self):
        with self.assertRaises(ParseError) as cm:
            parse(self.lex("3+"))
        self.assertEqual(
            str(cm.exception), "Missing operand(s) for operator '+' in expression '3+'"
        )

        with self.assertRaises(ParseError) as cm:
            parse(self.lex("(3+)"))
        self.assertEqual(
            str(cm.exception),
            "Missing operand(s) for operator '+' in expression '(3+)'",
        )

    def test_missing_operator(self):
        with self.assertRaises(ParseError) as cm:
            parse(self.lex("3 5"))
        self.assertEqual(
            str(cm.exception),
            "Unbalanced operators or operands in expression '3 5'",
        )

    def test_comment(self):
        tree = parse(self.lex("$(FOO) # comment\n"))
        self.assertEqual(str(tree), "$(FOO)")

    def test_unknown_token(self):
        with self.assertRaises(ParseError) as cm:
            parse(self.lex("3 +$"))
        self.assertEqual(
            str(cm.exception),
            "Unexpected component '$' from '3 +$' position 3",
        )

    def test_unexpected_operator(self):
        with self.assertRaises(ParseError) as cm:
            parse(self.lex("3 :6"))
        self.assertEqual(str(cm.exception), "Invalid expression '3 :6'")


example/dsc-parser.py
"""A demo script to parse DSC file and evaluate !if expression.
"""
import argparse
import enum
import logging
import re

import edk2_expression
import edk2_expression.error

logger = logging.getLogger("dsc-parser")


class Format(str, enum.Enum):
    Valid = "\033[1;32m+\033[0;32m "
    Invalid = "\033[1;31m-\033[0;31m "
    NotRelated = "\033[2;37m:\033[0m "
    Reset = "\033[0m"

    def __str__(self) -> str:
        return self.value


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("file", help="input file")
    parser.add_argument("-v", "--verbose", action="store_true", help="verbose output")
    args = parser.parse_args()

    logging.basicConfig(
        level=logging.DEBUG if args.verbose else logging.INFO,
        format="[%(levelname)s] %(message)s",
    )

    with open(args.file) as fd:
        file_content = fd.read()

    context = {}
    criteria = []
    for line in file_content.splitlines():
        # exclude comment and whitespaces
        clean_line = line.rsplit("#", 1)[0]
        clean_line = clean_line.strip()

        if not clean_line:
            continue

        # parse
        if m := re.fullmatch(r"DEFINE +(\w+) *= *(.+)", clean_line):
            name, raw_expr = m.groups()
            expr = edk2_expression.parse(raw_expr)
            logger.debug(f"Parse {name} = {expr!r}")

            if not criteria or criteria[-1]:
                context[name] = expr
                prefix = Format.Valid
            else:
                prefix = Format.Invalid
            print(f"{prefix}{line}{Format.Reset}")

        elif m := re.fullmatch(r"!(?:else)?if (.+)", clean_line):
            raw_expr = m.group(1)
            expr = edk2_expression.parse(raw_expr)

            if any(criteria):
                logger.debug(f"Stop evaluate !if {expr!r}")
                result = False
            else:
                result = bool(expr.evaluate(context, "evaluate"))
                logger.debug(f"Evaluate !if {expr!r} = {result}")

            criteria.append(result)
            prefix = Format.Valid if result else Format.Invalid
            print(f"{prefix}{line}{Format.Reset}")

        elif re.fullmatch(r"!else", clean_line):
            result = not any(criteria)
            criteria.append(result)

            prefix = Format.Valid if result else Format.Invalid
            print(f"{prefix}{line}{Format.Reset}")

        elif re.fullmatch(r"!endif", clean_line):
            print(f"{Format.Valid}{line}{Format.Reset}")
            criteria = []

        else:
            print(f"{Format.NotRelated}{line}{Format.Reset}")


if __name__ == "__main__":
    exit(main())


edk2_expression/ast/core.py
from __future__ import annotations

import enum
import typing
from abc import ABC, abstractmethod
from dataclasses import dataclass

from edk2_expression.error import NotSupported

if typing.TYPE_CHECKING:
    from pygments.token import Token


class NestMethod(str, enum.Enum):
    Error = "error"
    """Raise an error if a nested expression is encountered."""

    Ignore = "ignore"
    """Ignore nested expressions and return the original expression."""

    Evaluate = "evaluate"
    """Evaluate nested expressions and return the result."""

    @classmethod
    def _missing_(cls, value: str) -> NestMethod | None:
        name = value.lower()
        for member in cls:
            if member.value == name:
                return member


_DEFAULT_NEST_METHOD = NestMethod.Error


@dataclass(frozen=True)
class Expression(ABC):
    """Base class for all expressions."""

    @classmethod
    @abstractmethod
    def parse(cls, tokens: list[tuple[int, Token, str]]) -> Expression | None:
        """Parse a list of tokens into an expression AST. This method may pop
        items from the list when a expression is parsed.

        :param tokens: A list of tokens.
        :type tokens: list[tuple[int, Token, str]]
        :return: The parsed expression. The return value may be None if the
            input does not match format for this class.
        :rtype: Expression | None
        """

    @abstractmethod
    def __str__(self) -> str:
        """Return a string representation of the expression."""

    def evaluate(
        self, context: dict[str, object], nest: NestMethod | str = _DEFAULT_NEST_METHOD
    ) -> object:
        """Evaluate the expression and return the result.

        :param context: A dictionary of macro definitions.
        :type context: dict[str, object]
        :param nest: How to handle nested expressions. Use ``NestMethod.Ignore``
            ("ignore") to ignore nested expressions and return the original
            expression. Use ``NestMethod.Evaluate`` ("evaluate") to evaluate
            nested expressions and return the result. Default to ``NestMethod.Error``
            ("error") for raising an error if a nested expression is encountered.
        :type nest: NestMethod | str
        :return: The result of the evaluation.
        :rtype: object
        :raises NotSupported: If the expression cannot be evaluated.
        """
        raise NotSupported(
            f"Evaluation not supported for {type(self).__name__}: {self}"
        )


edk2_expression/lex.py
from __future__ import annotations

from pygments.lexer import RegexLexer, bygroups, include, words
from pygments.token import (
    Comment,
    Keyword,
    Name,
    Number,
    Operator,
    Punctuation,
    String,
    Whitespace,
)


class Edk2ExpressionLexer(RegexLexer):
    name = "EDK II Expression Syntax"
    aliases = []
    filenames = []
    mimetypes = []

    CCHAR = r'(?:[\x21\x23-\x26\x28-\x5b\x5d-\x7e\x20\x09]|\\[ntfrb0\\"\'])'
    CName = r"[a-zA-Z_][a-zA-Z0-9_]*"

    tokens = {
        "root": [
            include("whitespace"),
            (r"\n", Whitespace, "#pop"),
            (r"#.*", Comment.Single),
            (words(("{")), Punctuation, "array"),
            (words(("(")), Punctuation, "#push"),
            (words((")")), Punctuation, "#pop"),
            (
                words(
                    (
                        # fmt: off
                        "EQ", "NE", "LT", "GT", "LE", "GE", "AND", "and", "XOR"
                        "xor", "OR", "or", "NOT", "not",
                        # fmt: on
                    ),
                    prefix=r"\b",
                    suffix=r"\b",
                ),
                Operator,
            ),
            (
                words(
                    (
                        # fmt: off
                        "+", "-", "*", "/", "%", "<<", ">>", "==", "!=", "<", ">",
                        "<=", ">=", "|", "^", "&", "&&", "||", "!", "~", "?", ":"
                        # fmt: on
                    )
                ),
                Operator,
            ),
            (rf"\b{CName}\.{CName}\b", Name.Entity),  # <PcdName>
            (  # <RformatGuid>
                r"\b[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}\b",
                Name.Entity,
            ),
            include("constants"),
            (  # <MACROVAL>
                r"(\$\()([A-Z][A-Z0-9_]*)(\))",
                bygroups(Keyword.Declaration, Name.Variable, Keyword.Declaration),
            ),
            # (  # <MacroDefined>
            #     rf"\b(DEFINED)(\()({CName})(\))",
            #     bygroups(Keyword.Type, Punctuation, Name.Variable, Punctuation),
            # ),
            (rf"\b{CName}\b", Name.Variable),  # <CName>
        ],
        "array": [
            (words(("{")), Punctuation, "#push"),
            (words(("}", ")")), Punctuation, "#pop"),
            include("whitespace"),
            (words((",")), Punctuation),
            (  # <UintMac>
                r"\b(UINT(?:8|16|32|64))(\()(\d+)(\))",
                bygroups(Keyword.Type, Punctuation, Number.Integer, Punctuation),
            ),
            (  # <Label> & <Offset>
                rf"\b(LABEL|OFFSET_OF)(\()({CName})(\))",
                bygroups(Keyword.Type, Punctuation, Name.Variable, Punctuation),
            ),
            (  # <DevicePath>
                r'\b(DEVICE_PATH)(\(")([.]+?)("\))',
                bygroups(Keyword.Type, Punctuation, String, Punctuation),
            ),
            (  # <GuidStr> that uses <RformatGuid>
                r'\b(GUID)(\(")([a-fA-F0-9-]{36})("\))',
                bygroups(Keyword.Type, Punctuation, Name.Entity, Punctuation),
            ),
            (  # <GuidStr> that uses <CformatGuid>
                r"\b(GUID)(\()(\{)",
                bygroups(Keyword.Type, Punctuation, Punctuation),
                ("#push", "#push"),
            ),
            (  # <GuidStr> that uses <CName>
                rf"\b(GUID)(\()({CName})(\))",
                bygroups(Keyword.Type, Punctuation, Name.Variable, Punctuation),
            ),
            include("constants"),
        ],
        "constants": [
            (  # QuotedString
                rf'(L?")({CCHAR}*)(")',
                bygroups(Punctuation, String, Punctuation),
            ),
            (  # <SglQuotedString>
                rf"(L?')({CCHAR}*)(')",
                bygroups(Punctuation, String, Punctuation),
            ),
            (  # <TrueFalse>
                words(
                    ("TRUE", "True", "true", "FALSE", "False", "false"),
                    prefix=r"\b",
                    suffix=r"\b",
                ),
                Keyword.Constant,
            ),
            (r"\b0[xX][a-fA-F0-9]{2,8}\b", Number.Hex),  # <HexNumber>
            (r"\b\d+\b", Number.Integer),  # <Integer>
        ],
        "whitespace": [
            (r"[\x20\x09]+", Whitespace),
        ],
    }
